### PR TITLE
[CLOSES #17] BarberException has warnings + errors

### LINKED ISSUE #17: Split BarberException.problems into errors and warnings
This allows for configuring whether `allWarningsAsErrors` for stricter validation. Specifically in consideration of https://github.com/cashapp/barber/pull/16#discussion_r306838038

### PR DESCRIPTION
**Pull request by adrw**
* Split problems to warnings and errors
* Only errors in validation throw the exception
* New configuration for Barbershop to opt in to warnings as errors for stricter validation

### CODE REVIEW

File: barber/src/main/kotlin/app/cash/barber/BarbershopBuilder.kt [hunk 1]
```diff
@@ -55,53 +57,62 @@ class BarbershopBuilder : Barbershop.Builder {
...
     if (installedDocument.isEmpty()) {
-      problems.add("""
+      warnings.add("""
         |No Documents installed
       """.trimMargin())
     }
 
-    if (problems.isNotEmpty()) {
-      throw BarberException(problems)
+    // Warn if Documents are unused in an DocumentTemplate
```
Comment (hamdanjaveed) [inline barber/src/main/kotlin/app/cash/barber/BarbershopBuilder.kt]: Comment says `Warn if`, but the code instead adds an `error`. Change the comment to `Error if ...`?

File: barber/src/main/kotlin/app/cash/barber/BarbershopBuilder.kt [hunk 2]
```diff
@@ -55,53 +57,62 @@ class BarbershopBuilder : Barbershop.Builder {
...
-    if (!usedDocuments.containsAll(installedDocument)) {
-      val danglingDocuments = installedDocument.filter { document ->
-        !usedDocuments.contains(document)
-      }
-      problems.add("""
-      |Document installed that is not used in any installed DocumentTemplates
-      |$danglingDocuments
-      |
-    """.trimMargin())
+    if (warningsAsErrors && (errors.isNotEmpty() || warnings.isNotEmpty())) {
```
Comment (hamdanjaveed) [inline barber/src/main/kotlin/app/cash/barber/BarbershopBuilder.kt]: This `if` statement is wrong, if `warningsAsErrors == false` but `errors.isNotEmpty() == true`, the statement still evaluates to false and we'll never throw a `BarberException`.

You have a correct version of this `if` statement further down on `line 246`, I'd extract this statement out into a function like

```kotlin
// Not sure I like the name
fun throwIfErrorsFound(errors: List<String>) =
  if (errors.isNotEmpty() || (warnings.isNotEmpty() && warningsAsErrors)) {
    throw BarberException(errors = errors, warnings = warnings)
  }
```

Comment (wesleyk) [inline barber/src/main/kotlin/app/cash/barber/BarbershopBuilder.kt]: leave a note here why we exit early with errors at this point

File: barber/src/test/kotlin/app/cash/barber/BarberExceptionTest.kt [hunk 3]
```diff
@@ -32,7 +33,8 @@ class BarberExceptionTest {
       |Details
       |More Details
       |
-      |3) Charlie is unset
+      |Warnings
+      |1) Charlie is unset
```
Comment (wesleyk) [inline barber/src/test/kotlin/app/cash/barber/BarberExceptionTest.kt]: charlie!

File: barber/src/main/kotlin/app/cash/barber/BarbershopBuilder.kt (hunk 1 above)
Comment (adrw) [reply barber/src/main/kotlin/app/cash/barber/BarbershopBuilder.kt]: Thanks, this actually should be a warning, changed accordingly!

File: barber/src/main/kotlin/app/cash/barber/BarbershopBuilder.kt (hunk 2 above)
Comment (adrw) [reply barber/src/main/kotlin/app/cash/barber/BarbershopBuilder.kt]: Good suggestion (and catch on the faulty logic 😅). Pulled it out into a helper function 👍

**Pull request merged**